src/data/finsupp/lex.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.pi.lex
+import data.finsupp.order
+
+/-!
+# Lexicographic order on finitely supported functions
+
+This file defines the lexicographic order on `finsupp`.
+-/
+
+namespace finsupp
+
+variables {α N : Type*} [has_zero N]
+
+@[simp] lemma to_lex_apply (x : α →₀ N) (a : α) : to_lex x a = x a := rfl
+@[simp] lemma of_lex_apply (x : lex (α →₀ N)) (a : α) : of_lex x a = x a := rfl
+
```

Comment:
Sorry about this; I'd not considered that this wouldn't work. I think these lemmas shouldn't exist after all.
```suggestion
```


The reason we have these for `pi` is that lean will unfold semireducible definitions to discover whether they are pi types; meaning that `lex (Pi i, A i)` is considered a function type. It doesn't do the same thing to discover whether they have a `coe_to_fun` (so not for `finsupp`). Arguably the `pi` behavior is undesirable, but fixing it is almost certainly rather involved.

Reply:
Ok, thanks for the explanation!

I had not tried for too long to make them work: there are only so many places where you can insert a `lex, to_lex, of_lex`!  😀